### PR TITLE
Replace FileSourceParams path with URI

### PR DESCRIPTION
**Pull request by rdettai**
### Description

The `FileSourceParams` curently contains a path that is converted forth and back to a URI. This PR straitens up things to avoid these conversions.

### How was this PR tested?

Describe how you tested this PR.

### CODE REVIEW

File: quickwit/quickwit-config/src/source_config/mod.rs [hunk 1]
```diff
@@ -229,12 +228,18 @@ pub enum SourceParams {
...
 impl SourceParams {
-    pub fn file<P: AsRef<Path>>(filepath: P) -> Self {
-        Self::File(FileSourceParams::file(filepath))
+    pub fn file_from_uri(uri: Uri) -> Self {
+        Self::File(FileSourceParams {
+            filepath: Some(uri),
+        })
+    }
+
+    pub fn file_from_str<P: AsRef<str>>(filepath: P) -> anyhow::Result<Self> {
```
Comment (guilload) [inline quickwit/quickwit-config/src/source_config/mod.rs:237]: I don't think this is necessary, it's already super easy to go from `&str` to `Uri` and then to a file params object.

File: quickwit/quickwit-config/src/source_config/mod.rs [hunk 2]
```diff
@@ -249,37 +254,33 @@ pub struct FileSourceParams {
...
+    #[serde(deserialize_with = "uri_from_str")]
+    pub filepath: Option<Uri>, //< If None read from stdin.
 }
 
-/// Deserializing as an URI first to validate the input.
-///
-/// TODO: we might want to replace `PathBuf` with `Uri` directly in
-/// `FileSourceParams`
-fn absolute_filepath_from_str<'de, D>(deserializer: D) -> Result<Option<PathBuf>, D::Error>
+impl FromStr for FileSourceParams {
```
Comment (guilload) [inline quickwit/quickwit-config/src/source_config/mod.rs:261]: Same here.

File: quickwit/quickwit-lambda/src/indexer/model.rs [hunk 3]
```diff
@@ -58,14 +59,14 @@ mod tests {
         });
         let parsed_cust_event: IndexerEvent = serde_json::from_value(cust_event).unwrap();
         assert_eq!(
-            parsed_cust_event.uri(),
-            PathBuf::from("s3://quickwit-test/test.json"),
+            parsed_cust_event.uri().unwrap(),
+            Uri::from_str("s3://quickwit-test/test.json").unwrap(),
```
Comment (guilload) [inline quickwit/quickwit-lambda/src/indexer/model.rs:63]: I believe `Uri` implements `PartialEq` for `&str`.

**Review by guilload — APPROVED** [review]

**Comment by rdettai** [comment]
Closing this in favor of https://github.com/quickwit-oss/quickwit/pull/5148 that brings significant changes on top of the file source,

**Pull request closed without merging**